app/vmselect/prometheus/prometheus.go

```diff
@@ -695,16 +843,35 @@ func LabelsHandler(qt *querytracer.Tracer, startTime time.Time, at *auth.Token,
 		return err
 	}
 	denyPartialResponse := httputils.GetDenyPartialResponse(r)
-	sq := storage.NewSearchQuery(at.AccountID, at.ProjectID, cp.start, cp.end, cp.filterss, *maxLabelsAPISeries)
-	labels, isPartial, err := netstorage.LabelNames(qt, denyPartialResponse, sq, limit, cp.deadline)
-	if err != nil {
-		return fmt.Errorf("cannot obtain labels: %w", err)
+
+	rc := getResultsCollector()
+	defer putResultsCollector(rc)
+	if err = executeAcrossTenants(qt, at, cp, func(qt *querytracer.Tracer, at *auth.Token, _ bool) error {
```

Comment:
It's better to move executeAcrossTenants to netstorage level and add tenant related labelsl to the blocks parsing.

 It allows to aggregate metrics over tenants properly.